emcache/connection_pool.py

```diff
@@ -365,8 +365,16 @@ class WaitingForAConnectionContext(BaseConnectionContext):
     async def __aenter__(self) -> MemcacheAsciiProtocol:
         try:
             self._connection = await self._waiter
-        except asyncio.CancelledError:
-            self._connection_pool.remove_waiter(self._waiter)
+        except asyncio.CancelledError as exc:
+            try:
+                connection = self._waiter.result()
+            except asyncio.CancelledError:
+                # Waiter was cancelled before he could get a
+                # connection, we need to remove it
+                self._connection_pool.remove_waiter(self._waiter)
+            else:
+                # Waiter was cancelled right after wakeup. We need
+                # to release the connection.
+                self._connection_pool.release_connection(connection, exc=exc)
```

Comment:
Be careful if we provide an exception as a parameter this will be used here [1] which will result on closing the connection.

This is done in that way because Memcache does not support operation identification, so anytime that you have an issue at the network level you can not recycle the connection

Because this is a cancellation that happens on the client-side we can recycle the connection.

So just remove this parameter.

[1] https://github.com/pfreixes/emcache/pull/49/files#diff-ab7bf183d9513e4b587ce6c652d5ea3dca31e54224d0970690fc7f4ea1060ac5R298

Reply:
good point! thanks :)